### PR TITLE
Resized images

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
--------- edited by @davivel

**BLOCKED** by https://github.com/owncloud/core/issues/16250

--------- edited by @jabarros
TASKs:
- [X] Changes after CR @jabarros
- [X] [QA] Create 'resizedImages' branch in QA Repo  @jesmrec
- [X] [QA] Merge 'resizedImages' branch in QA Repo
- [X] [QA] Create test plan @jesmrec 
- [ ] [QA] Validate test plan @jesmrec [WIP]

bugs and improvements:
- [ ] Unsupported image formats https://github.com/owncloud/android/pull/1599#issuecomment-229400874
- [ ] Bad preview of landscape images https://github.com/owncloud/android/pull/1599#issuecomment-229406324 @davivel [WIP]

---

This PR is now against master.

As this PR introduces a rather complex feature I will summarize it here.
- Clicking on an image loads a reduced version to the cache.
- The width and height is determinated by the screen.
- the benefit is a very small, but yet usable preview of the image.
- it can be used to send via whatsapp, or as a file to another app (like k9) with the correct filename.
- resized images are not treated as downloaded files. If the user wants to have the real file, he has to long-click and select download.
- clicking on an image shows first a square thumbnail (if available) until the resized image is loaded

**ToDo**
- Zooming a resized image with at least 2x (which can get blurry)
  - show snackbar "automatically download full image on zoom: yes"
  - Clicking "yes" after second time will offer option "always" to prevent showing the question. 
  - time out the snackbar the second time by just waiting will show "never download full image" -> "yes".
  - after both show second snackbar without action "default behaviour can be changed in settings"
- In setting: new entry "click on image downloads ..."
  - full image to storage
  - resized image to cache

### CODE REVIEW

**Comment by tobiasKaminsky** [comment]
Snackbar is only allowing one action.
So I have adapted the behaviour.
@jancborchardt @AndyScherzinger @davivel

**Comment by tobiasKaminsky** [comment]
@jancborchardt @AndyScherzinger 
After thinking about it I think a snackbar (or similar) with two options would be better, or?

**Comment by jancborchardt** [comment]
I thought we said zooming should always download the image? Without showing any snackbar or something like that?

**Comment by tobiasKaminsky** [comment]
No, please read the other PR.
There me and other said that this against the expectation to save bandwidth and storage. So this should stay optional.
And with the snackbar approach every user can decide on its own how the default behaviour should be.

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky your approach works for me since you are right, the snackbar is defined with 0-1 action. For more than 1 action we would need to use dialogs or something else.

**Comment by davivel** [comment]
Having a look now, just read the comments in the thread.

First ideas:
  - zoom behaviour should be in a separate PR to limit the scope of testing.
  - about sending preview image to other apps instead of the full size image without the using being aware of it: will this behaviour fit the expectations of most of the users?

File: AndroidManifest.xml [hunk 1]
```diff
@@ -114,7 +114,7 @@
             android:name=".providers.FileContentProvider"
             android:authorities="@string/authority"
             android:enabled="true"
-            android:exported="false"
+            android:exported="true"
```
Comment (davivel) [inline AndroidManifest.xml]: Why is this needed in this PR?

File: AndroidManifest.xml [hunk 2]
```diff
@@ -125,6 +125,12 @@
             android:exported="false"
             android:label="@string/search_users_and_groups_hint" />
 
+        <provider
+            android:name=".ui.adapter.DiskLruImageCacheFileProvider"
```
Comment (davivel) [inline AndroidManifest.xml]: Why is a new exported provider needed?

File: src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java [hunk 3]
```diff
@@ -157,6 +165,12 @@ public ThumbnailGenerationTask(ImageView imageView, FileDataStorageManager stora
             mAccount = account;
         }
 
+        public ThumbnailGenerationTask(ImageView imageView, FileDataStorageManager storageManager,
+                                       Account account, ProgressBar progressWheel) {
+        this(imageView, storageManager, account);
```
Comment (davivel) [inline src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java]: Wrong indentention here.

**Comment by tobiasKaminsky** [comment]
> zoom behaviour should be in a separate PR to limit the scope of testing.

This is a good idea.

> about sending preview image to other apps instead of the full size image without the user being aware of it: will this behaviour fit the expectations of most of the users?

If the image is not downloaded I assume that users know that the resized version is used.
The user has to explicit download a file to use this.
But maybe we can show this to the user by using "send resized image".
And use "send original size" when downloaded.

File: AndroidManifest.xml (hunk 2 above)
Comment (tobiasKaminsky) [inline AndroidManifest.xml]: This is for sending cached images to other apps.
Technically only a content:// is sent which points to this provider, as far as I remember.

File: AndroidManifest.xml (hunk 1 above)
Comment (tobiasKaminsky) [inline AndroidManifest.xml]: I have no idea :/
A quick test shows no problems if I revert this.

File: src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java [hunk 4]
```diff
@@ -281,18 +329,25 @@ private Bitmap doOCFileInBackground() {
                             try {
                                 String uri = mClient.getBaseUri() + "" +
                                         "/index.php/apps/files/api/v1/thumbnail/" +
```
Comment (jabarros) [inline src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java]: Please, set a constant for this

File: src/com/owncloud/android/files/FileOperationsHelper.java [hunk 5]
```diff
@@ -324,6 +334,26 @@ public void sendDownloadedFile(OCFile file) {
         }
     }
 
+    public void sendCachedImage(OCFile file) {
+        if (file != null) {
+            Intent sendIntent = new Intent(android.content.Intent.ACTION_SEND);
+            // set MimeType
+            sendIntent.setType(file.getMimetype());
+//            sendIntent.putExtra(Intent.EXTRA_STREAM, Uri.parse("content://" + DiskLruImageCacheFileProvider.AUTHORITY + "/#" + file.getRemoteId() + "#" + file.getFileName()));
```
Comment (jabarros) [inline src/com/owncloud/android/files/FileOperationsHelper.java]: Remove this comment if it is code not needed

**Comment by jabarros** [comment]
@tobiasKaminsky In a short term, QA team is going to validate this functionality. Could you take a look to the CR changes and do a rebase on master?

**Comment by jesmrec** [comment]
The images in format .tiff are not previewed, because they are not supported (thumbnail is not shown). If you tap on a tiff image, a spinner appears forever on a black background. That can be confusing for the user, so that he/she does not know if there is a problem or not. If this image format (and another ones) are not supported, it would be better to show an alert or similar indicating that the image .tiff (or whatever) can not be previewed.

**Comment by jesmrec** [comment]
if you tap on a "landscape" image, is not correctly resized. 

This is the downloaded image:

screen shot 2016-06-29 at 17 49 24

The previewed (not dowloaded) in portrait:

screen shot 2016-06-29 at 17 48 41

The previewed (nor downloaded) in landscape:

screen shot 2016-06-29 at 17 58 01

In this case, this is a "landscape" image (longer than widther), so the preview should fit with the device landscape orientation. The device in portrait orientation should show something like the first attached file. 

cc @tobiasKaminsky @owncloud/android-developers

**Comment by jabarros** [comment]
@owncloud/android-developers 

due to high number of conflits detected during the rebase from master, I pushed the rebase with the fixes into a new one (resizedImages_fix_rebase) in order to avoid break this PR if some conflicts were not fixed okay.

So, if anyone will work here, please, better move to new resizedImages_fix_rebase branch.

**Comment by davivel** [comment]
Will resume where @jabarros left it, in the branch resizedImages_fix_rebase, fixing the second bug.

File: AndroidManifest.xml (hunk 2 above)
Comment (davivel) [inline AndroidManifest.xml]: That is a different functionality requiring specific tests. I will move it to a separate branch.

Please, include in the description of the PR a summary of all the changes. Otherwise the test plans could not match the scope of the changes, since we don't keep an strict order for the tasks of reviewing code and defining tests - tasks that should be independent.

Comment (davivel) [inline AndroidManifest.xml]: I'm sorry, you already commented it in the description: "it can be used to send via whatsapp, or as a file to another app".  

I apologize.

File: AndroidManifest.xml [hunk 6]
```diff
@@ -129,9 +129,9 @@
             android:name=".providers.FileContentProvider"
             android:authorities="@string/authority"
             android:enabled="true"
-            android:exported="true"
+            android:exported="false"
```
Comment (davivel) [inline AndroidManifest.xml]: Self-note: this was set to true with addition of Document Provider.

File: src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java [hunk 7]
```diff
@@ -336,24 +391,39 @@ private Bitmap handlePNG(Bitmap bitmap, int px){
             return resultBitmap;
         }
 
-        private Bitmap doFileInBackground() {
+        private Bitmap doFileInBackground(Boolean mIsThumbnail) {
```
Comment (davivel) [inline src/com/owncloud/android/datamodel/ThumbnailsCacheManager.java]: Self note: why not boolean?

File: AndroidManifest.xml [hunk 8]
```diff
@@ -152,6 +152,12 @@
             </intent-filter>
         </provider>
 
+        <provider
+            android:name=".ui.adapter.DiskLruImageCacheFileProvider"
+            android:authorities="com.owncloud.imageCache.provider"
```
Comment (davivel) [inline AndroidManifest.xml]: Self-note: check if should be in setup.xml

**Comment by davivel** [comment]
Rebased double checked and pushed here; no need to go to other branch.

**Comment by davivel** [comment]
> The images in format .tiff are not previewed, because they are not supported (thumbnail is not shown). If you tap on a tiff image, a spinner appears forever on a black background. That can be confusing for the user, so that he/she does not know if there is a problem or not. If this image format (and another ones) are not supported, it would be better to show an alert or similar indicating that the image .tiff (or whatever) can not be previewed.

Even more: if a resized preview can't be get, I'd say the file should be downloaded instead.

**Comment by davivel** [comment]
> In this case, this is a "landscape" image (longer than widther), so the preview should fit with the device landscape orientation. The device in portrait orientation should show something like the first attached file.

Previews should use the same rules to adapt the size of the image to the screen that those used to render the full image. In your sample the preview is resulting in a clip of the image, but it should be a rescaled version.

File: src/com/owncloud/android/ui/preview/PreviewImageFragment.java [hunk 9]
```diff
@@ -199,10 +205,56 @@ public void onSaveInstanceState(Bundle outState) {
...
         if (getFile() != null) {
-            mLoadBitmapTask = new LoadBitmapTask(mImageView, mMessageView, mProgressWheel);
-            //mLoadBitmapTask.execute(new String[]{getFile().getStoragePath()});
-//            mLoadBitmapTask.execute(getFile().getStoragePath());
-            mLoadBitmapTask.execute(getFile());
+            mImageView.setTag(getFile().getFileId());
+
+            if (mShowResizedImage){
+                Bitmap resizedImage = ThumbnailsCacheManager.getBitmapFromDiskCache(
+                        String.valueOf("r" + getFile().getRemoteId()));
```
Comment (davivel) [inline src/com/owncloud/android/ui/preview/PreviewImageFragment.java]: Prefix "r" in the identifier to request a resized preview, and prefix "t" in the identifier to request a thumbnail? Is not too tricky? Seems that ```ThumbnailsCacheManager``` needs different methods for both cases.

**Comment by davivel** [comment]
> Previews should use the same rules to adapt the size of the image to the screen that those used to render the full image. In your sample the preview is resulting in a clip of the image, but it should be a rescaled version.

And in the very end, the problem is what @tobiasKaminsky already stated in the very beginning:

> Remaining problem:
owncloud/core#16250

IMHO, this is **blocking**. 'Til that issue is not addressed, this shouldn't be merged.

**Comment by zgmnkv** [comment]
Hi guys, any update on this one?
Downloading all the images that I viewed consumes a lot of memory on my device.

**Comment by jesmrec** [comment]
not for the moment... you know, we are opened to contributions!!! :)

**Comment by jesmrec** [comment]
Will check with next core generation.
